tests/rust/tcp-test/close/mod.rs

```diff
@@ -171,3 +179,167 @@ fn close_connecting_socket(libos: &mut LibOS, remote: &SocketAddrV4) -> Result<(
+    match libos.wait(qt, Some(Duration::from_micros(0))) {
+        Ok(qresult) if qresult.qr_opcode == DEMI_OPC_CLOSE => {},
+        Ok(_) => anyhow::bail!("wait() should succeed with close"),
+        Err(_) => anyhow::bail!("wait() should succeed"),
+    }
+
+    Ok(())
+}
+
+/// Attempts to close a TCP socket that is bound.
+fn async_close_bound_socket(libos: &mut LibOS, local: &SocketAddrV4) -> Result<()> {
+    println!("{}", stringify!(async_close_bound_socket));
+
+    // Create a bound socket.
+    let sockqd: QDesc = libos.socket(AF_INET, SOCK_STREAM, 0)?;
+    libos.bind(sockqd, *local)?;
+
+    // Succeed to close socket.
+    let qt: QToken = libos.async_close(sockqd)?;
+
+    // Poll once to ensure the async_close() coroutine runs and finishes the close.
+    match libos.wait(qt, Some(Duration::from_micros(0))) {
+        Ok(qresult) if qresult.qr_opcode == DEMI_OPC_CLOSE => {},
+        Ok(_) => anyhow::bail!("wait() should succeed with close"),
+        Err(_) => anyhow::bail!("wait() should succeed"),
+    }
+
+    Ok(())
+}
+
+/// Attempts to close a TCP socket that is listening.
+fn async_close_listening_socket(libos: &mut LibOS, local: &SocketAddrV4) -> Result<()> {
+    println!("{}", stringify!(aysnc_close_listening_socket));
+
+    // Create a listening socket.
+    let sockqd: QDesc = libos.socket(AF_INET, SOCK_STREAM, 0)?;
+    libos.bind(sockqd, *local)?;
+    libos.listen(sockqd, 16)?;
+
+    // Succeed to close socket.
+    let qt: QToken = libos.async_close(sockqd)?;
+
+    // Poll once to ensure the async_close() coroutine runs and finishes the close.
+    match libos.wait(qt, Some(Duration::from_micros(0))) {
+        Ok(qresult) if qresult.qr_opcode == DEMI_OPC_CLOSE => {},
+        Ok(_) => anyhow::bail!("wait() should succeed with close"),
+        Err(_) => anyhow::bail!("wait() should succeed"),
+    }
+
+    Ok(())
+}
+
+/// Attempts to close a TCP socket that is accepting.
+fn async_close_accepting_socket(libos: &mut LibOS, local: &SocketAddrV4) -> Result<()> {
+    println!("{}", stringify!(async_close_accepting_socket));
+
+    // Create an accepting socket.
+    let sockqd: QDesc = libos.socket(AF_INET, SOCK_STREAM, 0)?;
+    libos.bind(sockqd, *local)?;
+    libos.listen(sockqd, 16)?;
+    let qt: QToken = libos.accept(sockqd)?;
+
+    // Poll once to ensure that the accept() co-routine runs.
+    match libos.wait(qt, Some(Duration::from_micros(0))) {
+        Err(e) if e.errno == libc::ETIMEDOUT => {},
+        Ok(_) => anyhow::bail!("wait() should not succeed"),
+        Err(_) => anyhow::bail!("wait() should timeout"),
+    }
+
+    // Succeed to close socket.
+    let qt_close: QToken = libos.async_close(sockqd)?;
+
+    // Poll once to ensure the async_close() coroutine runs and finishes the close.
+    match libos.wait(qt_close, Some(Duration::from_micros(0))) {
+        Ok(qresult) if qresult.qr_opcode == DEMI_OPC_CLOSE => {},
```

Comment:
Additionally, we must check `qresult.qr_ret == 0`.

Reply:
This should be zero, otherwise it should not be encoded as a `demi_opcode_t::DEMI_OPC_CLOSE`.